crates/ide_assists/src/handlers/generate_is_empty_from_len.rs

```diff
@@ -0,0 +1,225 @@
+use hir::{AssocItem, HasSource, Impl};
+use syntax::{
+    ast::{self, NameOwner},
+    AstNode, TextRange,
+};
+
+use crate::{
+    assist_context::{AssistContext, Assists},
+    AssistId, AssistKind,
+};
+
+// Assist: generate_is_empty_from_len
+//
+// Generates is_empty implementation from the len method.
+//
+// ```
+// impl MyStruct {
+//     p$0ub fn len(&self) -> usize {
+//         self.data.len()
+//     }
+// }
+// ```
+// ->
+// ```
+// impl MyStruct {
+//     pub fn len(&self) -> usize {
+//         self.data.len()
+//     }
+//
+//     pub fn is_empty(&self) -> bool {
+//         self.len() == 0
+//     }
+// }
+// ```
+pub(crate) fn generate_is_empty_from_len(acc: &mut Assists, ctx: &AssistContext) -> Option<()> {
+    let fn_node = ctx.find_node_at_offset::<ast::Fn>()?;
+    let fn_name = fn_node.name()?;
+
+    if fn_name.text() != "len" {
+        cov_mark::hit!(len_function_not_present);
+        return None;
+    }
+
+    if fn_node.param_list()?.params().next().is_some() {
+        cov_mark::hit!(len_function_with_parameters);
+        return None;
+    }
+
+    let impl_ = fn_node.syntax().ancestors().into_iter().find_map(ast::Impl::cast)?;
+    let impl_def = ctx.sema.to_def(&impl_)?;
+    if is_empty_implemented(ctx, &impl_def) {
+        cov_mark::hit!(is_empty_already_implemented);
+        return None;
+    }
+
+    let range = get_text_range_of_len_function(ctx, &impl_def)?;
+
+    acc.add(
+        AssistId("generate_is_empty_from_len", AssistKind::Generate),
+        "Generate a is_empty impl from a len function",
+        range,
+        |builder| {
+            let code = get_is_empty_code();
+            builder.insert(range.end(), code)
+        },
+    )
+}
+
+fn get_function_from_impl(ctx: &AssistContext, impl_def: &Impl, name: &str) -> Option<AssocItem> {
+    let db = ctx.sema.db;
+    impl_def.items(db).into_iter().filter(|item| matches!(item, AssocItem::Function(_value))).find(
+        |func| match func.name(db) {
+            Some(fn_name) => fn_name.to_string() == name,
+            None => false,
+        },
+    )
```

Comment:
This won't work if the function is defined in a different impl. We want to do it similar to how its done here with `iterate_method_candidates` https://github.com/rust-analyzer/rust-analyzer/blob/master/crates/ide_assists/src/handlers/replace_for_loop_with_for_each.rs#L73-L101

But in the callback we merely have to check the arguments and return type, something like
```rust
	let method = typ.iterate_method_candidates(
        sema.db,
        krate,
        &traits_in_scope,
        Some(&wanted_method),
        |_, func| {
            if func.ret_type(sema.db).is_bool() {
                return Some(func);
            }
            None
        },
    );
```

You will have to add `len` and `is_empty` as known names here to be able to use this https://github.com/Veykril/rust-analyzer/blob/db6364fecc2b0e9a95d9aaece820a86265b4cb4f\crates\hir_expand\src\name.rs#L190-L193